src/Network/HTTP/Affjax.purs

```diff
@@ -0,0 +1,158 @@
+import Data.Either (Either(..))
+import Data.Foreign (Foreign(..), F())
+import Data.Function (Fn4(), runFn4)
+import Data.Options (Option(), Options(), options, (:=), opt)
+import Network.HTTP.Affjax.Request
+import Network.HTTP.Affjax.Response
+import Network.HTTP.Affjax.ResponseType
+import Network.HTTP.Method (Method(..))
+import Network.HTTP.RequestHeader (RequestHeader())
+import Network.HTTP.ResponseHeader (ResponseHeader(), responseHeader)
+import Network.HTTP.StatusCode (StatusCode())
+
+-- | The effect type for AJAX requests made with Affjax.
+foreign import data Ajax :: !
+
+-- | The type for Affjax requests.
+type Affjax e a = Aff (ajax :: Ajax | e) (AffjaxResponse a)
+
+-- | Options type for Affjax requests.
+foreign import data AffjaxOptions :: *
+
+-- | The type of records that will be received as an Affjax response.
+type AffjaxResponse a =
+  { status :: StatusCode
+  , headers :: [ResponseHeader]
+  , response :: a
+  }
+
+-- | Sets the URL for a request.
+url :: Option AffjaxOptions String
+url = opt "url"
+
+-- | Sets the HTTP method for a request.
+method :: Option AffjaxOptions Method
+method = opt "method"
+
+-- | Sets the content to send in a request.
+content :: Option AffjaxOptions RequestContent
+content = opt "content"
+
+-- | Sets the headers to send with a request.
+headers :: Option AffjaxOptions [RequestHeader]
+headers = opt "headers"
+
+-- | Sets the HTTP auth username to send with a request.
+username :: Option AffjaxOptions String
+username = opt "username"
+
+-- | Sets the HTTP auth password to send with a request.
+password :: Option AffjaxOptions String
+password = opt "password"
+
+-- | Sets the expected response type for a request. This is not exposed outside
+-- | of the module as the `ResponseType` is set based on the `Responsable`
+-- | instance for the expected result content type.
+responseType = opt "responseType" :: Option AffjaxOptions ResponseType
+
+-- | Runs a request.
+affjax :: forall e a. Responsable a ->
+                      Options AffjaxOptions ->
+                      Affjax e a
+affjax r = makeAff <<< affjax' r
+
+-- | Runs a request directly in Eff.
+affjax' :: forall e a. Responsable a ->
+                       Options AffjaxOptions ->
+                       (Error -> Eff (ajax :: Ajax | e) Unit) ->
+                       (AffjaxResponse a -> Eff (ajax :: Ajax | e) Unit) ->
+                       Eff (ajax :: Ajax | e) Unit
+affjax' (Responsable read ty) opts eb cb =
+  runFn4 unsafeAjax responseHeader (options $ opts <> responseType := ty) eb cb'
+  where
+  cb' :: AffjaxResponse Foreign -> Eff (ajax :: Ajax | e) Unit
+  cb' res = case res { response = _  } <$> read res.response of
+    Left err -> eb $ error (show err)
+    Right res' -> cb res'
+
+get :: forall e a. Responsable a -> String -> Affjax e a
+get r addr = affjax r $ method := GET
+                     <> url := addr
+
+post :: forall e a. Responsable a -> String -> RequestContent -> Affjax e a
```

Comment:
The issue I had is it would give us orphan instances for `Blob`, `FormData`, `Document` and all the `ArrayView` types - they need `IsOption` instances. It also seems like `Responsable` should be `IsForeign a <= Responsable`, as `fromForeign` is just `read` from `IsForeign`, but then we have another orhpan instance situation...

I might revisit the free monad version yet again, maybe there is a way this can be made to work.